### PR TITLE
stricter buffers in SparseMatrixCSC

### PR DESCRIPTION
**Pull request by abraunst**
This is a rebase of #30676, without the addition of a capacity-querying function.

The general idea is about making `length(nonzeros(A))==nnz(A)` in `SparseMatrixCSC`, with strict buffer checking on exported functions (a few internal functions still return illegal buffers because the construction is done in several places, in that cases we create an empty matrix and resize the buffers afterwards. They should be clearly marked by comments)    

Adresses #30662. See also: #26560, #30435.

Still WIP.

### CODE REVIEW

**Comment by abraunst** [comment]
rebasing

**Comment by abraunst** [comment]
Test errors seem unrelated, rebasing.

**Comment by abraunst** [comment]
I'd say the failing check is unrelated. Could we run nanosoldier to look for performance regressions?

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/40523#issuecomment-823151145):
```
NanosoldierError: failed to run benchmarks against comparison commit: failed process: Process(`sudo /run/media/system/data/nanosoldier/cset/bin/cset shield -e su nanosoldier -- -c /run/media/system/data/nanosoldier/workdir/jl_CZM7oH/benchscript.sh`, ProcessExited(1)) [1]

```
Logs and partial data can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/421a07f_vs_d998c7e/report.md)
cc @christopher-dG

**Comment by abraunst** [comment]
Is he on strike?

**Comment by abraunst** [comment]
can we uhh... retry?

**Comment by KristofferC** [comment]
Looking at https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/421a07f_vs_d998c7e/logs/d998c7e74c84bb6f78916e71bf4efab936347170_against.out it seems it gets stuck at 

```
loading group "sort"... done (took 0.30277937 seconds)
loading group "sparse"... 
```

That's a bit suspicious since this code touches the sparse stuff.

**Comment by abraunst** [comment]
> Looking at https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/421a07f_vs_d998c7e/logs/d998c7e74c84bb6f78916e71bf4efab936347170_against.out it seems it gets stuck at
> 
> ```
> loading group "sort"... done (took 0.30277937 seconds)
> loading group "sparse"... 
> ```
> 
> That's a bit suspicious since this code touches the sparse stuff.

ah ... I'll have a look at that

**Comment by KristofferC** [comment]
I tried it locally and it had no problem...

```
loading group "scalar"... done (took 32.864596103 seconds)
loading group "sort"... done (took 2.51484769 seconds)
loading group "sparse"... done (took 9.814697974 seconds)
loading group "collection"... done (took 13.754587869 seconds)
```

Let's try again then 

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/40523#issuecomment-824582237):
```
NanosoldierError: failed to run benchmarks against comparison commit: failed process: Process(`sudo /run/media/system/data/nanosoldier/cset/bin/cset shield -e su nanosoldier -- -c /run/media/system/data/nanosoldier/workdir/jl_2biAVI/benchscript.sh`, ProcessExited(1)) [1]

```
Logs and partial data can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/1846f67_vs_7fe05db/report.md)
cc @christopher-dG

**Comment by vtjnash** [comment]
Ah, sorry, master is broken right now ("comparison commit") because we need this PR.

@nanosoldier `runbenchmarks("sparse" || "problem" || "broadcast" || "linalg", vs="@671bccba27e69b40dabe73409df2b1feba25944c")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/40523#issuecomment-824864491) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/109504d_vs_671bccb/report.md). cc @christopher-dG

**Comment by abraunst** [comment]
If I understand correctly the only things that do not seem noise are 2 memory regressions in broadcast `["broadcast", "sparse", ("(1000, 1000)", 1)]` and `["broadcast", "sparse", ("(10000000,)", 1)]` (possibly fault of the PR) and one of performance `["sparse", "index", ("spmat", "col", "array", 1000)]`, but this seems a fluke (it is just getindex, that should not be affected)

**Comment by vtjnash** [comment]
Yeah, LGTM

File: stdlib/SuiteSparse/src/cholmod.jl [hunk 1]
```diff
@@ -1970,4 +1957,58 @@ end
 (*)(A::SparseVecOrMat{Float64,Ti},
     B::Hermitian{Float64,SparseMatrixCSC{Float64,Ti}}) where {Ti} = sparse(Sparse(A)*Sparse(B))
 
+# Sort all the indices in each column for the construction of a CSC sparse matrix
+# sortBuffers!(A, sortindices = :sortcols)        # Sort each column with sort()
+function sortBuffers!(m, n, colptr::Vector{Ti}, rowval::Vector{Ti}, nzval::Vector{Tv}) where {Ti <: Integer, Tv}
```
Comment (KristofferC) [inline stdlib/SuiteSparse/src/cholmod.jl:1962]: Not, but this looks a bit non-standard to me, any special reason for the camel case?

Comment (abraunst) [reply stdlib/SuiteSparse/src/cholmod.jl:1962]: You're right, this comes from `sortSparseMatrixCSC` that was in sparsematrix.jl (which I didn't even remove for the moment), I'll rename it (it should not be exported though).

**Comment by vtjnash** [comment]
I think this can be merged, once the reviewer comments are addressed. Any objections?

**Comment by abraunst** [comment]
the memory regression should not be there, I'll investigate...

**Comment by abraunst** [comment]
Should be fixed. The problem was that by removing the `capacity` function from the PR, we lose track of what is the target length of the buffers (which were `sizehint`ed appropriately, but the allocated length was inaccessible).  So I resorted to returning invalid buffers from `_allocres` (which is internal). Can we do another round of benchmarking?

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks("sparse" || "problem" || "broadcast" || "linalg", vs="@671bccba27e69b40dabe73409df2b1feba25944c")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/40523#issuecomment-825591188) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/535b576_vs_671bccb/report.md). cc @christopher-dG

**Pull request merged**